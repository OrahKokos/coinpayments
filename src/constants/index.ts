export const API_VERSION = 1;
export const API_PROTOCOL = 'https:';
export const API_HOST = 'www.coinpayments.net';
export const API_PATH = '/api.php';
export const API_FORMAT = 'json';
export const API_VALID_RESPONSE = 'ok';

export const CMDS = Object.freeze({
  RATES: 'rates',
  GET_BASIC_INFO: 'get_basic_info',
  CREATE_TRANSACTION: 'create_transaction',
  GET_CALLBACK_ADDRESS: 'get_callback_address',
  GET_TX: 'get_tx_info',
  GET_TX_MULTI: 'get_tx_info_multi',
  GET_TX_LIST: 'get_tx_ids',
  BALANCES: 'balances',
  GET_DEPOSIT_ADDRESS: 'get_deposit_address',
  CREATE_TRANSFER: 'create_transfer',
  CREATE_WITHDRAWAL: 'create_withdrawal',
  CREATE_MASS_WITHDRAWAL: 'create_mass_withdrawal',
  CANCEL_WITHDRAWAL: 'cancel_withdrawal',
  CONVERT: 'convert',
  CONVERT_LIMITS: 'convert_limits',
  GET_WITHDRAWAL_HISTORY: 'get_withdrawal_history',
  GET_WITHDRAWAL_INFO: 'get_withdrawal_info',
  GET_CONVERSATION_INFO: 'get_conversion_info',
  GET_TAG_INFO: 'get_pbn_info',
  GET_TAG_LIST: 'get_pbn_list',
  BUY_TAG: 'buy_pbn_tags',
  CLAIM_TAG: 'claim_pbn_tag',
  UPDATE_TAG: 'update_pbn_tag',
  RENEW_TAG: 'renew_pbn_tag',
  DELETE_TAG: 'delete_pbn_tag',
  CLAIM_COUPON: 'claim_pbn_coupon',
});
