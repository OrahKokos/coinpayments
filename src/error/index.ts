export default class CoinpaymentsError extends Error {
  public name: string;
  public message: string;
  public extra: Record<string, unknown>;

  constructor(message: string, extra: Record<string, unknown> = {}) {
    super(message);
    Error.captureStackTrace(this, CoinpaymentsError);
    this.message = message;
    this.extra = extra;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
