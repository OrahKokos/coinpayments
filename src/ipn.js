const qs     = require('querystring');
const crypto = require('crypto');

module.exports = (function ()
{
	function IPN({merchantId, merchantSecret})
	{
		if(!merchantId || !merchantSecret)
		{
			throw "Merchant ID and Merchant Secret are needed";
		}

		function validateHMAC(parameters) 
		{
			let signature, paramString;
			paramString = qs.stringify(parameters).replace(/%20/g, '+')
			signature = crypto.createHmac('sha512', merchantSecret).update(paramString).digest('hex');
			return signature;
		}

		let methods = {
			validate: (ProvidedHMAC, ipnParams) =>
			{
				return new Promise((resolve, reject) => {

					if(!ProvidedHMAC || !ipnParams || !ipnParams.ipn_mode || ipnParams.ipn_mode != 'hmac' || merchantId != ipnParams.merchant)
					{
						return resolve({success: false, message: "COINPAYMENTS_INVALID_REQUEST"});
					}

					let ComputedHMAC = this.validateHMAC(ipnParams);
					if(ComputedHMAC != ProvidedHMAC)
					{
						return reject({success: false, message: "COINPAYMENTS_INVALID_REQUEST"});
					}

					if(ipnParams.status < 0)
					{
						return resolve({success: false, ipn: ipnParams, failed: true, message: "COINPAYMENTS_TRANSACTION_FAILED"});
					}

					if(ipnParams.status < 100)
					{
						return resolve({success: false, ipn: ipnParams, pending: true, message: "COINPAYMENTS_TRANSACTION_PENDING"});
					}

					if(ipnParams.status == 100)
					{
						return resolve({success: true, ipn: ipnParams});
					}
					return reject({success: false, message: "COINPAYMENTS_UNKOWN_REQUEST"});
				});
			},
			middleware: (req, res, next) =>
			{
				methods.verify(req.get("HMAC"), req.body).then( result => {
					if(result.success)
					{
						this.emit('ipn_complete', req.body);
					}
					
					if(result.pending)
					{
						this.emit('ipn_pending', req.body);
					}

					if(result.failed)
					{
						this.emit('ipn_fail', req.body);
					}
					return next();
				}).catch( error => {
					return next(error.message);
				});
			}
		};
		return methods;
	}
	return IPN;
})();