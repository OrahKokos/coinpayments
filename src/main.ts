import Coinpayments from './index'

export default Coinpayments
export { Coinpayments }

module.exports = Coinpayments
