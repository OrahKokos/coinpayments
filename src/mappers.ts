import { CoinpaymentsRequest } from './types/base'
import {
  CoinpaymentsCreateMassWithdrawalOpts,
  CoinpaymentsGetTxMultiOpts,
} from './types/options'

export const mapPayload = <ExpectedOptions>(
  opts: ExpectedOptions,
  defaultRequestFields: CoinpaymentsRequest
): CoinpaymentsRequest => {
  return {
    ...defaultRequestFields,
    ...opts,
  }
}

export const mapMassWithdrawalPayload = (
  withdrawalArray: CoinpaymentsCreateMassWithdrawalOpts,
  defaultFields: CoinpaymentsRequest
): CoinpaymentsRequest => {
  const payload = withdrawalArray.reduce((ops, w, index) => {
    ops[`wd[wd${index + 1}][amount]`] = w.amount
    ops[`wd[wd${index + 1}][address]`] = w.address
    ops[`wd[wd${index + 1}][currency]`] = w.currency

    if (w.dest_tag) {
      ops[`wd[wd${index + 1}][dest_tag]`] = w.dest_tag
    }

    return ops
  }, {})
  return {
    ...defaultFields,
    ...payload,
  }
}

export const mapGetTxMultiPayload = (
  txArray: CoinpaymentsGetTxMultiOpts,
  defaultFields: CoinpaymentsRequest
): CoinpaymentsRequest => {
  const payload = { txid: txArray.join('|') }
  return {
    ...defaultFields,
    ...payload,
  }
}
