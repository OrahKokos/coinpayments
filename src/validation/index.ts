import CoinpaymentsError from '../error';
import Schema from './schema';

import { CMDS } from '../constants';

import { CoinpaymentsRequest } from '../types/base';
import {
  CoinpaymentsCreateMassWithdrawalOpts,
  CoinpaymentsCreateWithdrawalOpts,
} from 'types/options';

export const getVariations = (validationSchema: any): Array<Array<string>> => {
  const variations = validationSchema.find(key => Array.isArray(key));
  const staticKeys = validationSchema.filter(key => !Array.isArray(key));
  if (!variations?.length) {
    return [validationSchema];
  }
  return variations.map(variation => {
    return [].concat(staticKeys).concat([variation]);
  });
};

export const getValidationSchema = (
  cmd: string,
): Array<Array<string> | string> => {
  if (!cmd) {
    throw new CoinpaymentsError('No method passed');
  }
  const validationSchema = Schema[cmd];
  if (!validationSchema) {
    throw new CoinpaymentsError(`No such method ${cmd}`);
  }
  return validationSchema;
};

export const validate = (
  validationSchema: any,
  options: CoinpaymentsRequest,
): boolean => {
  const optionKeys = Object.keys(options);
  const validationRules = getVariations(validationSchema);
  return validationRules.some(rule => rule.every(r => optionKeys.includes(r)));
};

export const validateMassWithDrawal = (
  options: CoinpaymentsRequest,
): boolean => {
  const regex =
    /(wd\[wd[0-9]*\]\[(amount|address|currency|dest_tag|domain|pbntag|add_tx_fee|currency2|ipn_url|auto_confirm|note)\]|cmd)/;
  return Object.keys(options).every(key => regex.test(key));
};

export const validatePayload = (options: CoinpaymentsRequest): void => {
  if (options.cmd === CMDS.CREATE_MASS_WITHDRAWAL) {
    if (!validateMassWithDrawal(options)) {
      throw new CoinpaymentsError('Invalid mass withdrawal payload', options);
    }
  }
  const validationSchema = getValidationSchema(options.cmd);
  if (!validate(validationSchema, options)) {
    throw new CoinpaymentsError('Missing parameters', options);
  }
};

export const hasOneOfFields = <T extends Record<PropertyKey, unknown>>(
  obj: T,
  keys: Array<string>,
): boolean => keys.filter(key => obj.hasOwnProperty(key)).length === 1;

export const validateMassWithdrawalOpts = (
  withdrawalOpts: CoinpaymentsCreateWithdrawalOpts,
) =>
  withdrawalOpts.currency &&
  withdrawalOpts.amount &&
  hasOneOfFields(withdrawalOpts, ['address', 'domain', 'pbntag']);

export const filterMassWithdrawalOpts = (
  massWithdrawalOpts: CoinpaymentsCreateMassWithdrawalOpts,
): CoinpaymentsCreateMassWithdrawalOpts =>
  massWithdrawalOpts.filter(validateMassWithdrawalOpts);
