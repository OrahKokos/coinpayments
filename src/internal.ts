import { createHmac } from 'crypto';
import { request as httpsRequest } from 'https';
import { stringify } from 'querystring';

import {
  API_HOST,
  API_PATH,
  API_PROTOCOL,
  API_VERSION,
  API_FORMAT,
  API_VALID_RESPONSE,
} from './constants';
import CoinpaymentsError from './error';
import { validatePayload } from './validation';

import {
  CoinpaymentsCredentials,
  CoinpaymentsRequest,
  CoinpaymentsInternalRequestOps,
  CoinpaymentsInternalResponse,
  CoinpaymentsReturnCallback,
} from './types/base';

export const getPrivateHeaders = (
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
) => {
  const { secret } = credentials;

  const paramString = stringify(options);
  const signature = createHmac('sha512', secret)
    .update(paramString)
    .digest('hex');

  return {
    'Content-Type': 'application/x-www-form-urlencoded',
    HMAC: signature,
  };
};

export const makeRequest = <ExpectedResponse>(
  reqOps: CoinpaymentsInternalRequestOps,
  options: CoinpaymentsRequest,
): Promise<ExpectedResponse> => {
  return new Promise((resolve, reject) => {
    const req = httpsRequest(reqOps, res => {
      let chunks = '';

      res.setEncoding('utf8');

      res.on('data', chunk => {
        chunks += chunk;
      });

      res.on('end', () => {
        let data: CoinpaymentsInternalResponse<ExpectedResponse> = {
          error: API_VALID_RESPONSE,
        };
        try {
          data = JSON.parse(chunks);
        } catch (e) {
          return reject(
            new CoinpaymentsError('Invalid response', { data: chunks }),
          );
        }

        if (data.error !== API_VALID_RESPONSE) {
          return reject(new CoinpaymentsError(data.error, { data }));
        }
        return resolve(data.result);
      });
    });
    req.on('error', reject);
    req.write(stringify(options));
    return req.end();
  });
};

export const getRequestOptions = (
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
): CoinpaymentsInternalRequestOps => {
  return {
    protocol: API_PROTOCOL,
    method: 'post',
    host: API_HOST,
    path: API_PATH,
    headers: getPrivateHeaders(credentials, options),
  };
};

export const applyDefaultOptionValues = (
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
): CoinpaymentsRequest => {
  return {
    ...options,
    version: API_VERSION,
    format: API_FORMAT,
    key: credentials.key,
  };
};

export const request = async <ExpectedResponse>(
  credentials: CoinpaymentsCredentials,
  options: CoinpaymentsRequest,
  callback?: CoinpaymentsReturnCallback<ExpectedResponse>,
): Promise<ExpectedResponse> => {
  try {
    validatePayload(options);
    options = applyDefaultOptionValues(credentials, options);
    const reqOps = getRequestOptions(credentials, options);
    const response: ExpectedResponse = await makeRequest<ExpectedResponse>(
      reqOps,
      options,
    );
    if (callback) {
      return callback(null, response);
    }
    return response;
  } catch (e) {
    if (callback) {
      return callback(e);
    }
    return Promise.reject(e);
  }
};
