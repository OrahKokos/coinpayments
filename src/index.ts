import { CMDS } from './constants';
import CoinpaymentsError from './error';
import {
  mapPayload,
  mapMassWithdrawalPayload,
  mapGetTxMultiPayload,
} from './mappers';
import { request } from './internal';

import {
  CoinpaymentsCredentials,
  CoinpaymentsReturnCallback,
} from './types/base';
import {
  CoinpaymentsRatesOpts,
  CoinpaymentsGetBasicInfoOpts,
  CoinpaymentsGetCallbackAddressOpts,
  CoinpaymentsCreateTransactionOpts,
  CoinpaymentsGetTxOpts,
  CoinpaymentsGetTxMultiOpts,
  CoinpaymentsGetTxListOpts,
  CoinpaymentsBalancesOpts,
  CoinpaymentsGetDepositAddressOpts,
  CoinpaymentsCreateMassWithdrawalOpts,
  CoinpaymentsConvertCoinsOpts,
  CoinpaymentsConvertLimitsOpts,
  CoinpaymentsGetWithdrawalHistoryOpts,
  CoinpaymentsGetWithdrawalInfoOpts,
  CoinpaymentsCancelWithdrawalOpts,
  CoinpaymentsGetConversionInfoOpts,
  CoinpaymentsGetProfileOpts,
  CoinpaymentsTagListOpts,
  CoinpaymentsClaimTagOpts,
  CoinpaymentsUpdateTagProfileOpts,
  CoinpaymentsRenewTagOpts,
  CoinpaymentsDeleteTagOpts,
  CoinpaymentsClaimCouponOpts,
  CoinpaymentsBuyTagOpts,
  CoinpaymentsCreateTransferOpts,
  CoinpaymentsCreateWithdrawalOpts,
} from './types/options';

import {
  CoinpaymentsRatesResponse,
  CoinpaymentsCreateTransactionResponse,
  CoinpaymentsBalancesResponse,
  CoinpaymentsCreateWithdrawalResponse,
  CoinpaymentsCreateMassWithdrawalResponse,
  CoinpaymentsGetTxResponse,
  CoinpaymentsGetWithdrawalInfoResponse,
  CoinpaymentsGetTxListResponse,
  CoinpaymentsGetTxMultiResponse,
  CoinpaymentsGetBasicInfoResponse,
  CoinpaymentsGetDepositAddressResponse,
  CoinpaymentsGetCallbackAddressResponse,
  CoinpaymentsCreateTransferResponse,
  CoinpaymentsConvertCoinsResponse,
  CoinpaymentsConvertLimitsResponse,
  CoinpaymentsGetWithdrawalHistoryResponse,
  CoinpaymentsConversionInfoResponse,
  CoinpaymentsGetProfileResponse,
  CoinpaymentsTagListResponse,
  CoinpaymentsUpdateTagProfileResponse,
  CoinpaymentsClaimTagResponse,
  CoinpaymentsRenameTagResponse,
  CoinpaymentsDeleteTagResponse,
  CoinpaymentsClaimCouponResponse,
  CoinpaymentsBuyTagResponse,
} from './types/response';
import { filterMassWithdrawalOpts } from './validation';

class Coinpayments {
  private credentials: CoinpaymentsCredentials;

  constructor({ key = '', secret = '' }: CoinpaymentsCredentials) {
    if (!key) {
      throw new CoinpaymentsError('Missing public key');
    }
    if (!secret) {
      throw new CoinpaymentsError('Missing private key');
    }

    this.credentials = { key, secret };

    this.getBasicInfo = this.getBasicInfo.bind(this);
    this.rates = this.rates.bind(this);
    this.balances = this.balances.bind(this);
    this.getDepositAddress = this.getDepositAddress.bind(this);
    this.createTransaction = this.createTransaction.bind(this);
    this.getCallbackAddress = this.getCallbackAddress.bind(this);
    this.getTx = this.getTx.bind(this);
    this.getTxList = this.getTxList.bind(this);
    this.getTxMulti = this.getTxMulti.bind(this);
    this.createTransfer = this.createTransfer.bind(this);
    this.convertCoins = this.convertCoins.bind(this);
    this.convertLimits = this.convertLimits.bind(this);
    this.getWithdrawalHistory = this.getWithdrawalHistory.bind(this);
    this.getWithdrawalInfo = this.getWithdrawalInfo.bind(this);
    this.getConversionInfo = this.getConversionInfo.bind(this);
    this.getProfile = this.getProfile.bind(this);
    this.tagList = this.tagList.bind(this);
    this.updateTagProfile = this.updateTagProfile.bind(this);
    this.claimTag = this.claimTag.bind(this);
    this.cancelWithdrawal = this.cancelWithdrawal.bind(this);
    this.createWithdrawal = this.createWithdrawal.bind(this);
    this.createMassWithdrawal = this.createMassWithdrawal.bind(this);
  }

  public rates(
    options?:
      | CoinpaymentsRatesOpts
      | CoinpaymentsReturnCallback<CoinpaymentsRatesResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsRatesResponse>,
  ): Promise<CoinpaymentsRatesResponse> {
    if (!options && !callback) {
      options = {};
    }
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    const requestPayload = mapPayload<CoinpaymentsRatesOpts>(options, {
      cmd: CMDS.RATES,
    });

    return request<CoinpaymentsRatesResponse>(
      this.credentials,
      requestPayload,
      callback,
    );
  }

  public createTransaction(
    options: CoinpaymentsCreateTransactionOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateTransactionResponse>,
  ) {
    return request<CoinpaymentsCreateTransactionResponse>(
      this.credentials,
      mapPayload<CoinpaymentsCreateTransactionOpts>(options, {
        cmd: CMDS.CREATE_TRANSACTION,
      }),
      callback,
    );
  }

  public balances(
    options?:
      | CoinpaymentsBalancesOpts
      | CoinpaymentsReturnCallback<CoinpaymentsBalancesResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsBalancesResponse>,
  ) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    return request<CoinpaymentsBalancesResponse>(
      this.credentials,
      mapPayload<CoinpaymentsBalancesOpts>(options, {
        cmd: CMDS.BALANCES,
      }),
      callback,
    );
  }

  public createWithdrawal(
    options: CoinpaymentsCreateWithdrawalOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateWithdrawalResponse>,
  ) {
    options = { auto_confirm: 1, ...options };
    return request<CoinpaymentsCreateWithdrawalResponse>(
      this.credentials,
      mapPayload<CoinpaymentsCreateWithdrawalOpts>(options, {
        cmd: CMDS.CREATE_WITHDRAWAL,
      }),
      callback,
    );
  }

  public createMassWithdrawal(
    withdrawalArray: CoinpaymentsCreateMassWithdrawalOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateMassWithdrawalResponse>,
  ) {
    return request<CoinpaymentsCreateMassWithdrawalResponse>(
      this.credentials,
      mapMassWithdrawalPayload(filterMassWithdrawalOpts(withdrawalArray), {
        cmd: CMDS.CREATE_MASS_WITHDRAWAL,
      }),
      callback,
    );
  }

  public cancelWithdrawal(
    options: CoinpaymentsCancelWithdrawalOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCancelWithdrawalOpts>,
  ) {
    return request<CoinpaymentsCancelWithdrawalOpts>(
      this.credentials,
      mapPayload<CoinpaymentsCancelWithdrawalOpts>(options, {
        cmd: CMDS.CANCEL_WITHDRAWAL,
      }),
      callback,
    );
  }

  public getTx(
    options: CoinpaymentsGetTxOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetTxResponse>,
  ) {
    return request<CoinpaymentsGetTxResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetTxOpts>(options, {
        cmd: CMDS.GET_TX,
      }),
      callback,
    );
  }

  public getWithdrawalInfo(
    options: CoinpaymentsGetWithdrawalInfoOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetWithdrawalInfoResponse>,
  ) {
    return request<CoinpaymentsGetWithdrawalInfoResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetWithdrawalInfoOpts>(options, {
        cmd: CMDS.GET_WITHDRAWAL_INFO,
      }),
      callback,
    );
  }

  public getTxMulti(
    txIdArray: CoinpaymentsGetTxMultiOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetTxMultiResponse>,
  ) {
    if (!(txIdArray instanceof Array) || !txIdArray.length) {
      const error = new CoinpaymentsError('Invalid argument', { txIdArray });
      if (callback) {
        return callback(error);
      }
      return Promise.reject(error);
    }
    return request<CoinpaymentsGetTxMultiResponse>(
      this.credentials,
      mapGetTxMultiPayload(txIdArray, {
        cmd: CMDS.GET_TX_MULTI,
      }),
      callback,
    );
  }

  public getTxList(
    options?:
      | CoinpaymentsGetTxListOpts
      | CoinpaymentsReturnCallback<CoinpaymentsGetTxListResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetTxListResponse>,
  ) {
    if (!options && !callback) {
      options = {};
    }
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    return request<CoinpaymentsGetTxListResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetTxListOpts>(options, {
        cmd: CMDS.GET_TX_LIST,
      }),
      callback,
    );
  }

  public getBasicInfo(
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetBasicInfoResponse>,
  ) {
    return request<CoinpaymentsGetBasicInfoResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetBasicInfoOpts>(
        {},
        {
          cmd: CMDS.GET_BASIC_INFO,
        },
      ),
      callback,
    );
  }

  public getDepositAddress(
    options: CoinpaymentsGetDepositAddressOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetDepositAddressResponse>,
  ) {
    return request<CoinpaymentsGetDepositAddressResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetDepositAddressOpts>(options, {
        cmd: CMDS.GET_DEPOSIT_ADDRESS,
      }),
      callback,
    );
  }

  public getCallbackAddress(
    options: CoinpaymentsGetCallbackAddressOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetCallbackAddressResponse>,
  ) {
    return request<CoinpaymentsGetCallbackAddressResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetCallbackAddressOpts>(options, {
        cmd: CMDS.GET_CALLBACK_ADDRESS,
      }),
      callback,
    );
  }

  public createTransfer(
    options: CoinpaymentsCreateTransferOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsCreateTransferResponse>,
  ) {
    return request<CoinpaymentsCreateTransferResponse>(
      this.credentials,
      mapPayload<CoinpaymentsCreateTransferOpts>(options, {
        cmd: CMDS.CREATE_TRANSFER,
        auto_confirm: true,
      }),
      callback,
    );
  }

  public convertCoins(
    options: CoinpaymentsConvertCoinsOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsConvertCoinsResponse>,
  ) {
    return request<CoinpaymentsConvertCoinsResponse>(
      this.credentials,
      mapPayload<CoinpaymentsConvertCoinsOpts>(options, {
        cmd: CMDS.CONVERT,
      }),
      callback,
    );
  }

  public convertLimits(
    options: CoinpaymentsConvertLimitsOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsConvertLimitsResponse>,
  ) {
    return request<CoinpaymentsConvertLimitsResponse>(
      this.credentials,
      mapPayload<CoinpaymentsConvertLimitsOpts>(options, {
        cmd: CMDS.CONVERT_LIMITS,
      }),
      callback,
    );
  }

  public getWithdrawalHistory(
    options?:
      | CoinpaymentsGetWithdrawalHistoryOpts
      | CoinpaymentsReturnCallback<CoinpaymentsGetWithdrawalHistoryResponse>,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetWithdrawalHistoryResponse>,
  ) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    return request<CoinpaymentsGetWithdrawalHistoryResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetWithdrawalHistoryOpts>(options, {
        cmd: CMDS.GET_WITHDRAWAL_HISTORY,
      }),
      callback,
    );
  }

  public getConversionInfo(
    options: CoinpaymentsGetConversionInfoOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsConversionInfoResponse>,
  ) {
    return request<CoinpaymentsConversionInfoResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetConversionInfoOpts>(options, {
        cmd: CMDS.GET_CONVERSATION_INFO,
      }),
      callback,
    );
  }

  public getProfile(
    options: CoinpaymentsGetProfileOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsGetProfileResponse>,
  ) {
    return request<CoinpaymentsGetProfileResponse>(
      this.credentials,
      mapPayload<CoinpaymentsGetProfileOpts>(options, {
        cmd: CMDS.GET_TAG_INFO,
      }),
      callback,
    );
  }

  public tagList(
    callback?: CoinpaymentsReturnCallback<CoinpaymentsTagListResponse>,
  ) {
    return request<CoinpaymentsTagListResponse>(
      this.credentials,
      mapPayload<CoinpaymentsTagListOpts>(
        {},
        {
          cmd: CMDS.GET_TAG_LIST,
        },
      ),
      callback,
    );
  }

  public updateTagProfile(
    options: CoinpaymentsUpdateTagProfileOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsUpdateTagProfileResponse>,
  ) {
    return request<CoinpaymentsUpdateTagProfileResponse>(
      this.credentials,
      mapPayload<CoinpaymentsUpdateTagProfileOpts>(options, {
        cmd: CMDS.UPDATE_TAG,
      }),
      callback,
    );
  }

  public claimTag(
    options: CoinpaymentsClaimTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsClaimTagResponse>,
  ) {
    return request<CoinpaymentsClaimTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsClaimTagOpts>(options, {
        cmd: CMDS.CLAIM_TAG,
      }),
      callback,
    );
  }

  public renewTag(
    options: CoinpaymentsRenewTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsRenameTagResponse>,
  ) {
    return request<CoinpaymentsRenameTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsRenewTagOpts>(options, {
        cmd: CMDS.RENEW_TAG,
      }),
      callback,
    );
  }

  public deleteTag(
    options: CoinpaymentsDeleteTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsDeleteTagResponse>,
  ) {
    return request<CoinpaymentsDeleteTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsDeleteTagOpts>(options, {
        cmd: CMDS.DELETE_TAG,
      }),
      callback,
    );
  }

  public claimCoupon(
    options: CoinpaymentsClaimCouponOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsClaimCouponResponse>,
  ) {
    return request<CoinpaymentsClaimCouponResponse>(
      this.credentials,
      mapPayload<CoinpaymentsClaimCouponOpts>(options, {
        cmd: CMDS.CLAIM_COUPON,
      }),
      callback,
    );
  }

  public buyTag(
    options: CoinpaymentsBuyTagOpts,
    callback?: CoinpaymentsReturnCallback<CoinpaymentsBuyTagResponse>,
  ) {
    return request<CoinpaymentsBuyTagResponse>(
      this.credentials,
      mapPayload<CoinpaymentsBuyTagOpts>(options, {
        cmd: CMDS.BUY_TAG,
      }),
      callback,
    );
  }
}

module.exports = Coinpayments;
module.exports.default = Coinpayments;
module.exports.Coinpayments = Coinpayments;

export { Coinpayments };
export default Coinpayments;
