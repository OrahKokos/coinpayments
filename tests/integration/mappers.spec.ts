import {
  mapMassWithdrawalPayload,
  mapGetTxMultiPayload,
} from '../../src/mappers';

import { validateMassWithDrawal } from '../../src/validation';

describe('Mappers integration tests', () => {
  describe('mapMassWithdrawalPayload integration tests', () => {
    it('Should return proper payload', () => {
      const VALID_PAYLOAD = [
        { currency: 'x', address: 'y', amount: 1000 },
        { currency: 'p', address: 'q', amount: 2000, dest_tag: 'dest' },
      ];
      const VALID_DEFAULTS = {
        cmd: 'very-cmd',
      };
      const payload = mapMassWithdrawalPayload(VALID_PAYLOAD, VALID_DEFAULTS);
      expect(payload).toHaveProperty('cmd', VALID_DEFAULTS.cmd);
      expect(validateMassWithDrawal(payload)).toBeTruthy();
      expect(payload).toMatchSnapshot();
    });
  });
  describe('mapGetTxMultiPayload integration tests', () => {
    it('Should return proper payload', () => {
      const VALID_PAYLOAD = ['a', 'b', 'c'];
      const VALID_DEFAULTS = {
        cmd: 'very-cmd',
      };
      const payload = mapGetTxMultiPayload(VALID_PAYLOAD, VALID_DEFAULTS);
      expect(payload).toHaveProperty('txid', VALID_PAYLOAD.join('|'));
      expect(payload).toHaveProperty('cmd', VALID_DEFAULTS.cmd);
      expect(payload).toMatchSnapshot();
    });
  });
});
