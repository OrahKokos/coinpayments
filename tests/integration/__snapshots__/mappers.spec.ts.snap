// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mappers integration tests mapGetTxMultiPayload integration tests Should return proper payload 1`] = `
Object {
  "cmd": "very-cmd",
  "txid": "a|b|c",
}
`;

exports[`Mappers integration tests mapMassWithdrawalPayload integration tests Should return proper payload 1`] = `
Object {
  "cmd": "very-cmd",
  "wd[wd1][address]": "y",
  "wd[wd1][amount]": 1000,
  "wd[wd1][currency]": "x",
  "wd[wd2][address]": "q",
  "wd[wd2][amount]": 2000,
  "wd[wd2][currency]": "p",
  "wd[wd2][dest_tag]": "dest",
}
`;
